### PR TITLE
Update K2 component modes

### PR DESCRIPTION
**Pull request by apolyakov**
Rework K2 component modes. Previously we had two component modes: `oneshot`, `multshot`. In this PR I introduce following modes:

1. `k2-cli`: most simple mode. Semantically equivalent to regular `PHP` script;
2. `k2-server`: semantically equivalent to current `KPHP`-server;
3. `k2-oneshot`: K2 oneshot component;
4. `k2-multishot`: K2 multishot component.

### CODE REVIEW

File: runtime-light/component/component.cpp [hunk 1]
```diff
@@ -51,12 +137,6 @@ void ComponentState::process_platform_updates() noexcept {
         }
       } else { // update on incoming stream
         php_debug("got new incoming stream %" PRIu64, stream_d);
-        if (standard_stream_ != INVALID_PLATFORM_DESCRIPTOR) {
```
Comment (astrophysik) [inline runtime-light/component/component.cpp:54]: Is there any guarantee that `k2-oneshot` component will not get second incoming stream?

File: runtime-light/component/component.cpp [hunk 2]
```diff
@@ -6,18 +6,104 @@
...
+    standard_stream_ = co_await init_kphp_server_component();
+  }
+}
+
+template task_t<void> ComponentState::run_component_prologue<ComponentKind::CLI>();
+template task_t<void> ComponentState::run_component_prologue<ComponentKind::Server>();
+template task_t<void> ComponentState::run_component_prologue<ComponentKind::Oneshot>();
+template task_t<void> ComponentState::run_component_prologue<ComponentKind::Multishot>();
+
+task_t<void> ComponentState::run_component_epilogue() noexcept {
```
Comment (astrophysik) [inline runtime-light/component/component.cpp:94]: Why prologue is has compile time conditions. but epilogue has runtime?

File: runtime-light/component/component.cpp [hunk 3]
```diff
@@ -6,18 +6,104 @@
...
+  } else if (magic == K2_INVOKE_JOB_WORKER_MAGIC) {
+    php_error("not implemented");
+  } else {
+    php_error("server got unexpected type of request: 0x%x", magic);
+  }
+
+  co_return stream_d;
+}
+
+int32_t merge_output_buffers(ComponentState &component_ctx) noexcept {
```
Comment (astrophysik) [inline runtime-light/component/component.cpp:55]: I think this function should be in another place. For example, output-buffer.h.

File: runtime-light/component/component.cpp [hunk 4]
```diff
@@ -6,18 +6,104 @@
...
+#include "runtime-light/coroutine/task.h"
 #include "runtime-light/header.h"
 #include "runtime-light/scheduler/scheduler.h"
+#include "runtime-light/streams/streams.h"
 #include "runtime-light/utils/context.h"
+#include "runtime-light/utils/json-functions.h"
+
+namespace {
+
+constexpr uint32_t K2_INVOKE_HTTP_MAGIC = 0xd909efe8;
```
Comment (astrophysik) [inline runtime-light/component/component.cpp:24]: Maybe create one more file, that will contain init functions?

File: runtime-light/component/component.h [hunk 5]
```diff
@@ -82,8 +97,9 @@ struct ComponentState {
   RegexComponentState regex_component_context;
 
 private:
-  task_t<void> main_task;
```
[the remark this answers is not in the document]
Comment (apolyakov) [reply runtime-light/component/component.h:85]: To get it a naming similar to other private members of ComponentContext

File: runtime-light/component/component.cpp (hunk 1 above)
Comment (apolyakov) [reply runtime-light/component/component.cpp:54]: It's guaranteed by the K2 platform

File: runtime-light/component/component.cpp (hunk 2 above)
Comment (apolyakov) [reply runtime-light/component/component.cpp:94]: Prologue is only called by the compiler, that's why we can use compile-time expressions. At the same time, epilogue is called not only by the compiler, but, for example, by exit functions too.

File: runtime-light/component/component.cpp (hunk 4 above)
Comment (apolyakov) [reply runtime-light/component/component.cpp:24]: I'll do it in the next PR.

**Pull request merged**